### ISSUE TITLE
[Ubuntu] Cloud-init failing with KeyError: id0

### ISSUE THREAD

**Issue by MaxRink** [kubernetes-sigs/image-builder#712]
**What steps did you take and what happened:**
with new ubuntu builds i am seeing cloud-init failing to assign networking with `KeyError: id0`

```
[2021-10-06 17:03:44] Cloud-init v. 21.3-1-g6803368d-0ubuntu1~20.04.3 running 'init-local' at Wed, 06 Oct 2021 17:03:44 +0000. Up 2.82 seconds.
[2021-10-06 17:03:44] 2021-10-06 17:03:44,289 - util.py[WARNING]: failed stage init-local
[2021-10-06 17:03:44] failed run of stage init-local
[2021-10-06 17:03:44] ------------------------------------------------------------
[2021-10-06 17:03:44] Traceback (most recent call last):
[2021-10-06 17:03:44]   File "/usr/lib/python3/dist-packages/cloudinit/cmd/main.py", line 682, in status_wrapper
[2021-10-06 17:03:44]     ret = functor(name, args)
[2021-10-06 17:03:44]   File "/usr/lib/python3/dist-packages/cloudinit/cmd/main.py", line 391, in main_init
[2021-10-06 17:03:44]     init.apply_network_config(bring_up=bool(mode != sources.DSMODE_LOCAL))
[2021-10-06 17:03:44]   File "/usr/lib/python3/dist-packages/cloudinit/stages.py", line 824, in apply_network_config
[2021-10-06 17:03:44]     return self.distro.apply_network_config(
[2021-10-06 17:03:44]   File "/usr/lib/python3/dist-packages/cloudinit/distros/__init__.py", line 220, in apply_network_config
[2021-10-06 17:03:44]     network_state = parse_net_config_data(netconfig)
[2021-10-06 17:03:44]   File "/usr/lib/python3/dist-packages/cloudinit/net/network_state.py", line 1074, in parse_net_config_data
[2021-10-06 17:03:44]     nsi.parse_config(skip_broken=skip_broken)
[2021-10-06 17:03:44]   File "/usr/lib/python3/dist-packages/cloudinit/net/network_state.py", line 261, in parse_config
[2021-10-06 17:03:44]     self.parse_config_v2(skip_broken=skip_broken)
[2021-10-06 17:03:44]   File "/usr/lib/python3/dist-packages/cloudinit/net/network_state.py", line 310, in parse_config_v2
[2021-10-06 17:03:44]     self._v2_common(command)
[2021-10-06 17:03:44]   File "/usr/lib/python3/dist-packages/cloudinit/net/network_state.py", line 722, in _v2_common
[2021-10-06 17:03:44]     self._handle_individual_nameserver(name_cmd, iface)
[2021-10-06 17:03:44]   File "/usr/lib/python3/dist-packages/cloudinit/net/network_state.py", line 91, in decorator
[2021-10-06 17:03:44]     return func(self, command, *args, **kwargs)
[2021-10-06 17:03:44]   File "/usr/lib/python3/dist-packages/cloudinit/net/network_state.py", line 546, in _handle_individual_nameserver
[2021-10-06 17:03:44]     _iface[iface]['dns'] = {'nameservers': nameservers, 'search': search}
[2021-10-06 17:03:44] KeyError: 'id0'
[2021-10-06 17:03:44] ------------------------------------------------------------
[2021-10-06 17:03:44] Cloud-init v. 21.3-1-g6803368d-0ubuntu1~20.04.3 running 'init' at Wed, 06 Oct 2021 17:03:44 +0000. Up 3.19 seconds.
```

metadata:
```
instance-id: "wlan-1-md-0-775d8846bf-9bfrd"
local-hostname: "wlan-1-md-0-775d8846bf-9bfrd"
wait-on-network:
  ipv4: false
  ipv6: false
network:
  version: 2
  ethernets:
    id0:
      match:
        macaddress: "00:50:56:a1:d8:a7"
      set-name: "eth0"
      wakeonlan: true
      addresses:
      - "10.196.27.122/28"
      gateway4: "10.196.27.126"
      nameservers:
        addresses:
        - "10.102.102.132"
        - "10.102.102.133"
        - "10.90.24.1"
        search:
        - "refsa1.bn.schiff.telekom.de"

```

**What did you expect to happen:**
Cloud init is working properly

**Environment:**

Project: CAPI

Additional info for Image Builder for Cluster API related issues:

- OS (e.g. from `/etc/os-release`, or `cmd /c ver`): Ubuntu 2004

/kind bug
[One or more /area label. See https://github.com/kubernetes-sigs/cluster-api/labels?q=area for the list of labels]

**Comment by randomvariable** [comment]
The metadata looks correct, so not sure what's happening. @akutz has the most context about cloud-init bits.

**Comment by akutz** [comment]
I just posted this to the Cloud-Init IRC channel:

> Someone is hitting a weird error with the Cluster API VM images - https://github.com/kubernetes-sigs/image-builder/issues/712. Looks like a KeyError (more info in link) when parsing the network config for DNS. Except here is where it gets odd. The provided metadata matches the network v2 config example from https://cloudinit.readthedocs.io/en/latest/topics/network-config-format-v2.html, where "nameservers" is how you specify DNS info for a given interface. But the code, which hasn't changed in 5 years in this section, says the key should be "dns" - https://github.com/canonical/cloud-init/blame/main/cloudinit/net/network_state.py#L133-L146.

Basically, I agree, the following looks correct:

```yaml
instance-id: "wlan-1-md-0-775d8846bf-9bfrd"
local-hostname: "wlan-1-md-0-775d8846bf-9bfrd"
wait-on-network:
  ipv4: false
  ipv6: false
network:
  version: 2
  ethernets:
    id0:
      match:
        macaddress: "00:50:56:a1:d8:a7"
      set-name: "eth0"
      wakeonlan: true
      addresses:
      - "10.196.27.122/28"
      gateway4: "10.196.27.126"
      nameservers:
        addresses:
        - "10.102.102.132"
        - "10.102.102.133"
        - "10.90.24.1"
        search:
        - "refsa1.bn.schiff.telekom.de"
```

It's even what the [docs say](https://cloudinit.readthedocs.io/en/latest/topics/network-config-format-v2.html). But I'll be damned if [the code](https://github.com/canonical/cloud-init/blob/de166ec3e796c842330865d56c39962f0db45aac/cloudinit/net/network_state.py#L133-L146) for the stack trace, which is five years old, doesn't indicate that you are supposed to use `dns` instead of `nameservers`:

```python
    @property
    def dns_nameservers(self):
        try:
            return self._network_state['dns']['nameservers']
        except KeyError:
            return []


    @property
    def dns_searchdomains(self):
        try:
            return self._network_state['dns']['search']
        except KeyError:
            return []
```

I'm waiting on feedback from the cloud-init devs. I'll relay it here once I hear from them.

**Comment by MaxRink** [comment]
it got introduced somewhere between 20.1 and 203. 
Given that we use a patch release iso it actually might be caused by the new vmware datasource which is enabled for new installs.

**Comment by akutz** [comment]
I will look at this when I have time. The thing passes all the provided unit tests, so if this is a bug (which it appears to be), it's an odd one. I tested e2e by using Impish as well as upgrading a 21.04 install to Cloud-Init v21.3. I did not validate the older releases update channels.

**Comment by MaxRink** [comment]
hmm, we still use the 20.04.1-legacy iso and are nowhere pinning the version of cloud-init.
We have a few extra roles that for example install the HWE Kernel, maybe that brings v21.3 with it

**Comment by MaxRink** [comment]
or https://github.com/kubernetes-sigs/image-builder/blob/master/images/capi/ansible/roles/setup/tasks/debian.yml#L51, which does a dist upgrade.
That should also yield 20.04.3

**Comment by akutz** [comment]
Well there you go. All I know is what I see in your log from above :) 

```
2021-10-06 17:03:44] Cloud-init v. 21.3-1-g6803368d-0ubuntu1~20.04.3 running 'init-local' at Wed, 06 Oct 2021 17:03:44 +0000. Up 2.82 seconds.
```

That sure looks like Cloud-Init 21.3 ~ 20.04.3 to me.

Okay, well, based on your latest response, it seems y'all are upgrading to 20.04.3 and pulling in Cloud-Init 21.3 and **not** removing the old DS logic. That ain't good :)

**Comment by akutz** [comment]
Well crud, it's not the presence of both datasources. I just rebuilt the Ubuntu OVA with the following patch:

```diff
diff --git a/images/capi/ansible/roles/providers/tasks/vmware.yml b/images/capi/ansible/roles/providers/tasks/vmware.yml
index 06f88871..25143d93 100644
--- a/images/capi/ansible/roles/providers/tasks/vmware.yml
+++ b/images/capi/ansible/roles/providers/tasks/vmware.yml
@@ -47,17 +47,20 @@
     url:  '{{ guestinfo_datasource_script }}'
     dest: /tmp/cloud-init-vmware.sh
     mode: 0700
+  when: ansible_os_family != "Debian"
 
 - name: Execute cloud-init-vmware.sh
   shell: bash -o errexit -o pipefail /tmp/cloud-init-vmware.sh
   environment:
     REPO_SLUG: '{{ guestinfo_datasource_slug }}'
     GIT_REF:   '{{ guestinfo_datasource_ref }}'
+  when: ansible_os_family != "Debian"
 
 - name: Remove cloud-init-vmware.sh
   file:
     path:  /tmp/cloud-init-vmware.sh
     state: absent
+  when: ansible_os_family != "Debian"
 
 - name: Remove cloud-init /etc/cloud/cloud.cfg.d/99-disable-networking-config.cfg
   file:
```

And I can verify I get the same error, and that only the new DataSourceVMware is present:

![ScreenCapture-2021 10 07-17 17 46](https://user-images.githubusercontent.com/101085/136470039-d175610a-2a08-4927-bf8a-08fcafc161a2.png)
![ScreenCapture-2021 10 07-17 17 55](https://user-images.githubusercontent.com/101085/136470038-e3fd8fee-1dba-4fb7-8fa4-50bc95a5c0a6.png)
![ScreenCapture-2021 10 07-17 18 54](https://user-images.githubusercontent.com/101085/136470037-fc1623ca-380c-4629-bc75-2bbf9d9cb9c7.png)

**Comment by akutz** [comment]
> Now, I have no clue why the hostname is no longer set when fixing the above issue. It worked when the DNS issue occurred, but now it's not. I am just running cloud-init clean and rebooting the system in between attempts, so maybe that's why?

Ah, I believe the hostname isn't being set exactly *because* of the DNS being set correctly and the local IP unable to be resolved via that DNS server.

**Comment by akutz** [comment]
This is occurring because the current value of `_iface` has the renamed interface, not the original key from `iface`. This is confirmed with the addition of some logging:

```diff
diff --git a/cloudinit/net/network_state.py b/cloudinit/net/network_state.py
index 95b064f0..06ff8e96 100644
--- a/cloudinit/net/network_state.py
+++ b/cloudinit/net/network_state.py
@@ -543,7 +543,11 @@ class NetworkStateInterpreter(metaclass=CommandHandlerMeta):
     def _handle_individual_nameserver(self, command, iface):
         _iface = self._network_state.get('interfaces')
         nameservers, search = self._parse_dns(command)
-        _iface[iface]['dns'] = {'nameservers': nameservers, 'search': search}
+        try:
+            _iface[iface]['dns'] = {'nameservers': nameservers, 'search': search}
+        except:
+            print("original iface name: %s\niface dict: %s\n" % (iface, _iface))
+            raise
 
     @ensure_command_keys(['destination'])
     def handle_route(self, command):
diff --git a/cloudinit/net/tests/test_network_state.py b/cloudinit/net/tests/test_network_state.py
index 84e8308a..45e99171 100644
--- a/cloudinit/net/tests/test_network_state.py
+++ b/cloudinit/net/tests/test_network_state.py
@@ -52,6 +52,7 @@ network:
     eth1:
       match:
         macaddress: '66:77:88:99:00:11'
+      set-name: "ens92"
       nameservers:
         search: [foo.local, bar.local]
         addresses: [4.4.4.4]
```

This produces the following output when running the test:

```bash
original iface name: eth1
iface dict: {'eth0': {'name': 'eth0', 'type': 'physical', 'mac_address': '00:11:22:33:44:55', 'inet': 'inet', 'mode': 'manual', 'mtu': None, 'address': None, 'gateway': None, 'subnets': [], 'accept-ra': None, 'wakeonlan': None, 'dns': {'nameservers': ['8.8.8.8'], 'search': ['spam.local', 'eggs.local']}}, 'ens92': {'name': 'ens92', 'type': 'physical', 'mac_address': '66:77:88:99:00:11', 'inet': 'inet', 'mode': 'manual', 'mtu': None, 'address': None, 'gateway': None, 'subnets': [], 'accept-ra': None, 'wakeonlan': None}}
```

Notice the original value of `iface` is `eth1`, but the dict has the renamed `ens92`.

**Comment by akutz** [comment]
@MaxRink communicated to me on Kubernetes Slack that the above patch, once applied to his system, also fixes the issue. Max, I leave it up to you if you want to close this or leave it open. I suggest closing it and opening a new issue to pin Cloud-Init at v21.1 until there's an update to it in the Ubuntu `updates` channel.

**Comment by zawachte-msft** [comment]
Yeap ... sigh `set-name` is broken for the networkd renderer. https://bugs.launchpad.net/cloud-init/+bug/1949407

**Comment by akutz** [comment]
This new bug is not the same as the previous. A new Kube image builder issue should be created to track this. For example, https://github.com/kkeshavamurthy/image-builder/commit/fbb715256b7477700922d65e332d5002513a29f1 does not fix #712 as #712 is fixed. Again, this is a greenfield issue with an identical root cause.

**Comment by zawachte-msft** [comment]
Yes, I agree. We don't use image-builder (although, we consult the repo regularly) so I didn't know if any of the images built with this repo are using networkd or hitting this code path and wasn't sure if it was in scope to create an issue here.

**Comment by akutz** [comment]
> Yes, I agree. We don't use image-builder (although, we consult the repo regularly) so I didn't know if any of the images built with this repo are using networkd or hitting this code path and wasn't sure if it was in scope to create an issue here.

Hi @zawachte-msft,

Well, someone _would_ hit this issue had they used the networkd renderer. @MaxRink previously hit the custom DNS issue instead. As I indicated, they both have the same root cause -- they only occur when `set-name` is used. However, the fix is different for each and both bugs have different stack traces. Regardless, thanks for bringing it to our attention. I actually saw this pop up in Cloud-Init's repo a day or two before you added this comment and thought "I wonder how long until someone files something against the image-builder..." 😄
